federationapi/internal/perform.go

```diff
@@ -210,10 +210,16 @@ func (r *FederationInternalAPI) performJoinUsingServer(
 	}
 	r.statistics.ForServer(serverName).Success()
 
-	authEvents := respSendJoin.AuthEvents.UntrustedEvents(respMakeJoin.RoomVersion)
+	// If the remote server returned an event in the "event" key of
+	// the send_join request then we should use that instead. It may
+	// contain signatures that we don't know about.
+	if respSendJoin.Event != nil {
+		event = respSendJoin.Event
```

Comment:
Is this safe to do in the general case? Or can this be used to get homeservers to arbitrarily sign things?

Reply:
I'll add some checks here to ensure the membership event is well-formed and what we expect, but if it is, then yes it should be safe.